packages/mui-material/src/AppBar/AppBar.js

```diff
@@ -72,30 +76,61 @@ const AppBarRoot = styled(Paper, {
     ...(ownerState.position === 'relative' && {
       position: 'relative',
     }),
-    ...(ownerState.color === 'default' && {
-      backgroundColor: backgroundColorDefault,
-      color: theme.palette.getContrastText(backgroundColorDefault),
-    }),
-    ...(ownerState.color &&
-      ownerState.color !== 'default' &&
-      ownerState.color !== 'inherit' &&
-      ownerState.color !== 'transparent' && {
-        backgroundColor: theme.palette[ownerState.color].main,
-        color: theme.palette[ownerState.color].contrastText,
+    ...(!theme.vars && {
+      ...(ownerState.color === 'default' && {
+        backgroundColor: backgroundColorDefault,
+        color: theme.palette.getContrastText(backgroundColorDefault),
+      }),
+      ...(ownerState.color &&
+        ownerState.color !== 'default' &&
+        ownerState.color !== 'inherit' &&
+        ownerState.color !== 'transparent' && {
+          backgroundColor: theme.palette[ownerState.color].main,
+          color: theme.palette[ownerState.color].contrastText,
+        }),
+      ...(ownerState.color === 'inherit' && {
+        color: 'inherit',
+      }),
+      ...(theme.palette.mode === 'dark' &&
+        !ownerState.enableColorOnDark && {
+          backgroundColor: null,
+          color: null,
+        }),
+      ...(ownerState.color === 'transparent' && {
+        backgroundColor: 'transparent',
+        color: 'inherit',
+        ...(theme.palette.mode === 'dark' && {
+          backgroundImage: 'none',
+        }),
       }),
-    ...(ownerState.color === 'inherit' && {
-      color: 'inherit',
     }),
-    ...(theme.palette.mode === 'dark' &&
-      !ownerState.enableColorOnDark && {
-        backgroundColor: null,
-        color: null,
+    ...(theme.vars && {
+      ...(ownerState.color === 'default' && {
+        '--AppBar-background': ownerState.enableColorOnDark
+          ? theme.vars.palette.AppBar.defaultBg
+          : joinVars(theme.vars.palette.AppBar.darkBg, theme.vars.palette.AppBar.defaultBg),
+        '--AppBar-color': ownerState.enableColorOnDark
+          ? theme.vars.palette.text.primary
+          : joinVars(theme.vars.palette.AppBar.darkColor, theme.vars.palette.text.primary),
```

Comment:
@mnajdova the case of AppBar is tricky but I managed to make it works.

To prevent dark mode flickers, we need to make sure that the stylesheet does not change between modes. That's why I have to introduce:

```js
// component variables
--AppBar-background
--AppBar-color

// and use it as usual
backgroundColor: 'var(--AppBar-background)',
color: 'var(--AppBar-color)',
```

and then at the global theme tokens, create `--mui-palette-AppBar-darkBg` **only** for dark mode.

This will allow the AppBar to use the fallback value when it is on light mode without changing the component stylesheet.

```js
...(ownerState.color === 'default' && {
    '--AppBar-background': ownerState.enableColorOnDark
      ? theme.vars.palette.AppBar.defaultBg
      : joinVars(theme.vars.palette.AppBar.darkBg, theme.vars.palette.AppBar.defaultBg),
    '--AppBar-color': ownerState.enableColorOnDark
      ? theme.vars.palette.text.primary
      : joinVars(theme.vars.palette.AppBar.darkColor, theme.vars.palette.text.primary),
```

Reply:
This is closer to what I had in mind, the only difference that I was thinking about is setting the value of `unset`, but not generating the CSS vars is even better 👍

Reply:
> setting the value of unset

Setting the variable to `unset` would not work. `unset` in considered valid in CSS so it won't use the fallback value.

```css
// the background will be transparent
--variable-1: unset;
background-color: var(--variable-1, #fff);
```

```css
// the background will be white
background-color: var(--variable-1, #fff);
```